test/radix/quic_ops.c

```diff
@@ -0,0 +1,1055 @@
+}
+
+DEF_FUNC(hf_write_ex2)
+{
+    int ok = 0, r;
+    SSL *ssl;
+    const void *buf;
+    size_t buf_len, bytes_written = 0;
+    uint64_t flags;
+
+    F_POP(flags);
+    F_POP2(buf, buf_len);
+    REQUIRE_SSL(ssl);
+
+    r = SSL_write_ex2(ssl, buf, buf_len, flags, &bytes_written);
+    if (!TEST_true(r)
+        || !check_consistent_want(ssl, r)
+        || !TEST_size_t_eq(bytes_written, buf_len))
+        goto err;
+
+    ok = 1;
+err:
+    return ok;
+}
+
+DEF_FUNC(hf_write_fail)
+{
+    int ok = 0, ret;
+    SSL *ssl;
+    size_t bytes_written = 0;
+
+    REQUIRE_SSL(ssl);
+
+    ret = SSL_write_ex(ssl, "apple", 5, &bytes_written);
+    if (!TEST_false(ret)
+        || !TEST_true(check_consistent_want(ssl, ret))
+        || !TEST_size_t_eq(bytes_written, 0))
+        goto err;
+
+    ok = 1;
+err:
+    return ok;
+}
+
+DEF_FUNC(hf_read_expect)
+{
+    int ok = 0, r;
+    SSL *ssl;
+    const void *buf;
+    size_t buf_len, bytes_read = 0;
+
+    F_POP2(buf, buf_len);
+    REQUIRE_SSL(ssl);
+
+    if (buf_len > 0 && RT()->tmp_buf == NULL
+        && !TEST_ptr(RT()->tmp_buf = OPENSSL_malloc(buf_len)))
+        goto err;
+
+    r = SSL_read_ex(ssl, RT()->tmp_buf + RT()->tmp_buf_offset,
+                    buf_len - RT()->tmp_buf_offset,
+                    &bytes_read);
+    if (!TEST_true(check_consistent_want(ssl, r)))
+        goto err;
+
+    if (!r)
+        F_SPIN_AGAIN();
+
+    if (bytes_read + RT()->tmp_buf_offset != buf_len) {
+        RT()->tmp_buf_offset += bytes_read;
+        F_SPIN_AGAIN();
+    }
+
+    if (buf_len > 0
+        && !TEST_mem_eq(RT()->tmp_buf, buf_len, buf, buf_len))
+        goto err;
+
+    OPENSSL_free(RT()->tmp_buf);
+    RT()->tmp_buf         = NULL;
+    RT()->tmp_buf_offset  = 0;
+
+    ok = 1;
+err:
+    return ok;
+}
+
+DEF_FUNC(hf_read_fail)
+{
+    int ok = 0, r;
+    SSL *ssl;
+    char buf[1] = {0};
+    size_t bytes_read = 0;
+    uint64_t do_wait;
+
+    F_POP(do_wait);
+    REQUIRE_SSL(ssl);
+
+    r = SSL_read_ex(ssl, buf, sizeof(buf), &bytes_read);
+    if (!TEST_false(r)
+        || !TEST_true(check_consistent_want(ssl, r))
+        || !TEST_size_t_eq(bytes_read, 0))
+        goto err;
+
+    if (do_wait && is_want(ssl, 0))
+        F_SPIN_AGAIN();
+
+    ok = 1;
+err:
+    return ok;
+}
+
+DEF_FUNC(hf_connect_wait)
+{
+    int ok = 0, ret;
+    SSL *ssl;
+
+    REQUIRE_SSL(ssl);
+
+    /* if not started */
+    if (RT()->scratch0 == 0) {
+        if (!TEST_true(SSL_set_blocking_mode(ssl, 0)))
+            return 0;
+
+        /* 0 is the success case for SSL_set_alpn_protos(). */
+        if (!TEST_false(SSL_set_alpn_protos(ssl, alpn_ossltest,
+                                            sizeof(alpn_ossltest))))
+            goto err;
+    }
+
+    RT()->scratch0 = 1; /* connect started */
+    ret = SSL_connect(ssl);
+    radix_activate_slot(0);
+    if (!TEST_true(check_consistent_want(ssl, ret)))
+        goto err;
+
+    if (ret != 1) {
+        if (1 /* TODO */ && is_want(ssl, ret))
```

Comment:
TODO what? I would prefer to have at least decent hint to see what's missing here.